Improve assert failure reporting

lundibundi opened the pull request:
Parts of https://github.com/metarhia/metatests/pull/64 I wanted to have even without `util.inspect`.

This case:
```js
metatests.testSync('', (test) => test.strictSame('1', 1));
```
Was reported as:
```
 Test failed:

        Type: strictSame
       Stack: Error
            ImperativeTest.strictSame (/lib/imperative-test.js:194:10)
            ImperativeTest.metatests.testSync [as func] (/test/imperative.js:182:39)
            ImperativeTest.runNow (/lib/imperative-test.js:357:25)
            process.nextTick (/lib/imperative-test.js:346:12)
    Expected: 1
      Actual: 1
     Message: undefined
```

Now it is:
```
 Test failed:

        Type: strictSame
     Message: undefined
    Expected: 1 || number
      Actual: "1" || string
       Stack: Error
            ImperativeTest.strictSame (/lib/imperative-test.js:194:10)
            ImperativeTest.metatests.testSync [as func] (/test/imperative.js:183:39)
            ImperativeTest.runNow (/lib/imperative-test.js:357:25)
            process.nextTick (/lib/imperative-test.js:346:12)
```

belochub commented on lib/report.js:
```diff
@@ -112,7 +100,17 @@ class ConciseReporter extends Reporter {
+      if (key === 'actual' || key === 'expected') {
+        let type = typeof value;
+        if (type === 'object') type = value.constructor.name;
+        value = `${JSON.stringify(value)} || ${type}`;
```
I don't like using `'||'` to delimit value and type.
Using `':'` will probably be easier to understand.

lundibundi replied:
I just wanted to make a distinction between the value and type clear (i.e objects will contain colons as key-value separators). So now if you see || you may be sure that next comes the value type (unless obviously there is || in a string but that's much less possible).
Though I'm open to other separators.

belochub replied:
@lundibundi What do you think about using `is` as a separator, I think it will make it more human-readable,
e.g. `{} is Object`, `10 is number`, `'str' is string`?

lundibundi replied:
looks good, though not as distinct-looking as || but good enough. I hope it will be better when we add color support.

belochub commented on lib/report.js:
```diff
@@ -112,7 +100,17 @@ class ConciseReporter extends Reporter {
     this.printAssertErrorSeparator();
-    console.log(list(res, message, ['success']) + '\n');
+    let s = message;
+    for (const key of ['type', 'message', 'expected', 'actual', 'stack']) {
```
Wouldn't it be better to iterate `Object.keys(res)` here and just add a check
```javascript
    if (key === 'success') continue;
```
inside a loop?

lundibundi replied:
Nope, I needed specific key order. It was simple iteration before which resulted in unsighty output.

belochub replied:
And what about replacing this for loop with `forEach`?

lundibundi replied:
Okay, I'm fine either way.

belochub commented:
Should we also maybe skip message printing if it is not set?

tshemsedinov commented:
For `test.error(err);` we can also skip `Expected` and `Actual` and print just an error stack.

lundibundi commented:
@tshemsedinov hmm, I have actually been thinking of setting custom values in such cases. I.e 'No error' string as expected in 'error' case etc.
As for skipping, I think it would be better to skip stack and only set the 'actual' to make it clear of what the error is.

tshemsedinov commented:
@lundibundi If no message provided for `test.error(err);` we really can print default message, but if we explain that error is not expected here why do we need `Expected: null` and `Actual: Error message...`?

lundibundi commented:
I didn't suggest to print a message. I wanted to
* print `Expected: No Error` instead of `Expected: null`
* don't print `trace` because it's much more clear that we have `Expected` and `Actual` than just `trace` as having trace doesn't explain what the error was (you have to deduce that because this test is `error` it means that it must not have an error and therefore `trace` is actually the error that has not been expected. Which is in my opinion much less clear than having `Actual` and `Expected` for every test)

lundibundi commented:
PTAL @belochub @tshemsedinov. I addressed the comments, [this](https://github.com/metarhia/metatests/pull/77#issuecomment-417913176) is left for the follow-up PRs.

belochub commented on lib/report.js:
```diff
@@ -112,7 +100,17 @@ class ConciseReporter extends Reporter {
+      let value = res[key];
+      if (key === 'actual' || key === 'expected') {
+        let type = typeof value;
+        if (type === 'object') type = value.constructor.name;
```
There is also a `null`-check missing here, if the `value` is `null`, this line throws.

belochub commented on lib/report.js:
```diff
@@ -112,7 +100,18 @@ class ConciseReporter extends Reporter {
+      if (key === 'actual' || key === 'expected') {
+        let type = typeof value;
+        if (value === null) type = 'null';
+        else if (type === 'object') type = value.constructor.name;
```
To make it completely safe:
```javascript
if (type === 'object' && value !== null && value.constructor)
  type = value.constructor.name;
```

lundibundi commented:
Landed in 11b9a228805a11a99955287b1519deb63840a714.
